File: detection/src/predict_traffic.py
from fastapi import FastAPI
from pydantic import BaseModel
import joblib
import numpy as np
import uvicorn

# Load model and scaler
model = joblib.load("traffic_predictor.pkl")
scaler = joblib.load("traffic_scaler.pkl")

# Initialize FastAPI app
app = FastAPI()

# Define request body using Pydantic
class TrafficInput(BaseModel):
    hour: int
    dayofweek: int
    is_weekend: int
    temperature: float
    humidity: float
    rain: float
    wind_speed: float
    vehicle_count_last_5min: int
    weather_condition: int

@app.post("/predict")
def predict_traffic(data: TrafficInput):
    # Prepare features
    features = [
        data.hour,
        data.dayofweek,
        data.is_weekend,
        data.temperature,
        data.humidity,
        data.rain,
        data.wind_speed,
        data.vehicle_count_last_5min,
        data.weather_condition
    ]
    X = np.array(features).reshape(1, -1)
    X_scaled = scaler.transform(X)
    predicted_vehicle_count = model.predict(X_scaled)[0]

    return {"predicted_vehicle_count": round(float(predicted_vehicle_count), 2)}


if __name__ == "__main__":
    uvicorn.run("predict_traffic:app", host="0.0.0.0", port=8090, reload=False)

File: composeApp/src/desktopMain/kotlin/com/gub/features/analytics/data/PredictTraffic.kt
package com.gub.features.analytics.data

import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.engine.cio.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.request.*
import io.ktor.http.*
import io.ktor.serialization.kotlinx.json.json
import kotlinx.coroutines.*
import kotlinx.serialization.*
import kotlinx.serialization.json.*

@Serializable
data class TrafficInput(
    val hour: Int,
    val dayofweek: Int,
    val is_weekend: Int,
    val temperature: Double,
    val humidity: Double,
    val rain: Double,
    val wind_speed: Double,
    val vehicle_count_last_5min: Int,
    val weather_condition: Int
)

@Serializable
data class PredictionResponse(
    val predicted_vehicle_count: Double
)

val client = HttpClient(CIO) {
    install(ContentNegotiation) {
        json()
    }
}

fun predictTraffic() {
    val input = TrafficInput(
        hour = 17,
        dayofweek = 1,
        is_weekend = 0,
        temperature = 31.5,
        humidity = 72.0,
        rain = 0.1,
        wind_speed = 8.0,
        vehicle_count_last_5min = 50,
        weather_condition = 0
    )

    CoroutineScope(Dispatchers.IO).launch {
        try {
            val response: PredictionResponse = client.post("http://localhost:8090/predict") {
                contentType(ContentType.Application.Json)
                setBody(input)
            }.body()

            println("✅ Predicted Vehicle Count: ${response.predicted_vehicle_count}")
        } catch (e: Exception) {
            println("❌ API Call Failed: ${e.message}")
        }
    }
}